R/shift.R

```diff
@@ -0,0 +1,114 @@
+#' @export
+shift <- function(x, n = 1L, default = NULL, order_by = NULL) {
+  n <- as_scalar_integer(n)
+  shift_impl(x, n = n, default = default, order_by = NULL)
+}
+
+#' @export
+#' @rdname shift
+lag <- function(x, n = 1L, default = NULL, order_by = NULL) {
+  n <- as_positive_scalar_integer(n)
+  shift_impl(x, n, default, order_by)
+}
+
+#' @export
+#' @rdname shift
+lead <- function(x, n = 1L, default = NULL, order_by = NULL) {
+  n <- as_positive_scalar_integer(n)
+  shift_impl(x, n * -1L, default, order_by)
+}
+
+shift_impl <- function(x, n, default = NULL, order_by = NULL, error_call = caller_env()) {
+  size <- vec_size(x)
+
+  if (!is.null(order_by)) {
+    out <- with_order(x, order_by, size, shift, n = n, default = default, error_call = error_call)
+    return(out)
+  }
+
+  if (identical(n, 0L)) {
+    return(x)
+  }
+
+  lag <- sign(n) > 0L
+  n <- abs(n)
+
+  if (n > size) {
+    n <- size
+  }
+
+  if (is.null(default)) {
+    vec_shift_slice(x, n, size, lag, error_call = error_call)
+  } else {
+    vec_shift_c(x, n, size, lag, default, error_call = error_call)
+  }
+}
+
+as_scalar_integer <- function(n, arg_name = "n", error_call = caller_env()) {
+  withCallingHandlers({
+    vec_assert(n, size = 1L, arg = "n")
+    vec_cast(n, integer(), x_arg = "n")
+  }, error = function(cnd) {
+    msg <- "{.arg {arg_name}} must be a single positive number."
+    cli_abort(msg, parent = cnd, call = error_call)
+  })
+}
+
+as_positive_scalar_integer <- function(n, arg_name = "n", error_call = caller_env()) {
+  n <- as_scalar_integer(n, arg_name = arg_name, error_call = error_call)
+  if (n < 0L) {
+    msg <- "{.arg {arg_name}} must be positive."
+    cli_abort(msg, call = error_call)
+  }
+  n
+}
+
+vec_shift_slice <- function(x, n, size, lag, error_call = caller_env()) {
+  idx_default <- vec_rep(NA_integer_, n)
+
+  if (lag) {
+    idx <- seq2(1L, size - n)
+    idx <- c(idx_default, idx)
+    vec_slice(x, idx)
+  } else {
+    idx <- seq2(1L + n, size)
+    idx <- c(idx, idx_default)
+    vec_slice(x, idx)
+  }
+}
+
+vec_shift_c <- function(x, n, size, lag, default, error_call = caller_env()) {
+  withCallingHandlers({
```

Comment:
Probably won't need the `withCallingHandlers()` thing when `vec_assert()` and `vec_cast()` accepts a `call=`.

Reply:
Would something like this to instrument-add `.__error_call__.` into existing functions be useful ?

``` r
library(rlang)
library(vctrs)

hidden <- function(f) {
  b <- body(f)
  stopifnot(identical(b[[1]], sym("{")))
  
  new_function(
    formals(f), 
    as.call(new_node(node_car(b), new_node(quote(.__error_call__. <- caller_env()), node_cdr(b)))), 
    environment(f)
  )
  
}

vec_assert2 <- hidden(vctrs::vec_assert)

f <- function() {
  vec_assert2(1:10, size = 3)
}
g <- function() {
  vec_assert(1:10, size = 3)
}

tryCatch(f(), error = print)
#> <error/vctrs_error_assert_size>
#> Error in `f()`: `1:10` must have size 3, not size 10.
#> Backtrace:
#>  1. base::tryCatch(f(), error = print)
#>  5. global f()
#>  6. vctrs vec_assert2(1:10, size = 3)
tryCatch(g(), error = print)
#> <error/vctrs_error_assert_size>
#> Error in `vec_assert()`: `1:10` must have size 3, not size 10.
#> Backtrace:
#>  1. base::tryCatch(g(), error = print)
#>  5. global g()
#>  6. vctrs::vec_assert(1:10, size = 3)
```

<sup>Created on 2021-11-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>